src/main/java/com/cloudbees/plugins/credentials/CredentialsSelectHelper.java

```diff
@@ -608,8 +616,30 @@ public JSONObject doAddCredentials(StaplerRequest2 req, StaplerResponse2 rsp) th
                         .element("notificationType", "ERROR");
             }
             store.checkPermission(CredentialsStoreAction.CREATE);
-            Credentials credentials = Descriptor.bindJSON(req, Credentials.class, data.getJSONObject("credentials"));
-            boolean credentialsWereAdded = store.addCredentials(wrapper.getDomain(), credentials);
+            boolean credentialsWereAdded;
+            try {
+                Credentials credentials = Descriptor.bindJSON(req, Credentials.class,
+                                                              data.getJSONObject("credentials"));
+                credentialsWereAdded = store.addCredentials(wrapper.getDomain(), credentials);
+            } catch (LinkageError e) {
+                /*
+                 * Descriptor#newInstanceImpl throws a LinkageError if the constructor throws any exception other
+                 * than HTTP response exceptions such as Descriptor.FormException.
+                 *
+                 * This approach is taken to maintain backward compatibility, as throwing a FormException directly
+                 * from the constructor would result in a source-incompatible change, potentially breaking dependent plugins.
+                 *
+                 * Here, known exceptions are caught specifically to provide meaningful error response.
+                 */
+                Throwable rootCause = ExceptionUtils.getRootCause(e);
+                if (rootCause instanceof IOException || rootCause instanceof IllegalArgumentException
+                    || rootCause instanceof GeneralSecurityException) {
```

Comment:
Why the 3 exceptions? AFAICT only `IllegalArgumentException` will be the root exception, right?

Reply:
> Why the 3 exceptions? AFAICT only IllegalArgumentException will be the root exception, right?

The reason for handling IOException and GeneralSecurityException along side IllegalArgumentException is due to how exceptions are wrapped and propagated. In this case, the constructor itself catches IOException and GeneralSecurityException and rethrows them as IllegalArgumentException. However, when these exceptions bubble up to this catch block, they are wrapped at multiple levels, ultimately resulting in a LinkageError.
When we retrieve the root cause, it points back to the original IOException or GeneralSecurityException rather than the rethrown IllegalArgumentException.
https://github.com/jenkinsci/credentials-plugin/blob/master/src/main/java/com/cloudbees/plugins/credentials/impl/CertificateCredentialsImpl.java#L703
https://github.com/jenkinsci/credentials-plugin/blob/master/src/main/java/com/cloudbees/plugins/credentials/impl/CertificateCredentialsImpl.java#L142

> Also, is it possible in this plugin to see the LinkageError with a different root cause than the IllegalArgumentException? Why is the if needed?

Regarding LinkageError, the check for different root causes is a safeguard. Although IllegalArgumentException is the expected root cause in most scenarios, this ensures that other potential but unexpected root causes (like related to class loading or dependencies) are not overlooked and are handled appropriately.

Reply:
> In this case, the constructor itself catches IOException and GeneralSecurityException and rethrows them as IllegalArgumentException

but for those cases, the root cause of the LinkageError would be the IllegalArgumentException and not the IOException or the GeneralSecurityException. Or am I missing something?

Reply:
> > Also, is it possible in this plugin to see the LinkageError with a different root cause than the IllegalArgumentException? Why is the if needed?
> 
> Regarding LinkageError, the check for different root causes is a safeguard. Although IllegalArgumentException is the expected root cause in most scenarios, this ensures that other potential but unexpected root causes (like related to class loading or dependencies) are not overlooked and are handled appropriately.

fair enough

Reply:
> > In this case, the constructor itself catches IOException and GeneralSecurityException and rethrows them as IllegalArgumentException
> 
> but for those cases, the root cause of the LinkageError would be the IllegalArgumentException and not the IOException or the GeneralSecurityException. Or am I missing something?

The constructor catches IOException and GeneralSecurityException and rethrows them as an IllegalArgumentException. This means that IllegalArgumentException becomes the direct exception thrown, and the original exceptions (IOException or GeneralSecurityException) are set as its cause.

When LinkageError is thrown, it is wrapping the IllegalArgumentException. However, ExceptionUtils#getRootCause traverse the entire chain of exceptions and stop at the deepest cause.

In this case, the deepest cause (root cause) would still be the original IOException or GeneralSecurityException, not the rethrown IllegalArgumentException.